File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.js
function solution(orders, course) {

    const map = new Map();

    function combination(arr, target, idx, str) {

        if (str.length === target) {

            map.set(str, (map.get(str) || 0) + 1);

            return;
        }

        for (let i = idx; i < arr.length; i++) {
            combination(arr, target, i + 1, str + arr[i]);
        }
    }

    // 모든 조합 생성
    for (const order of orders) {

        const sorted = [...order].sort();

        for (const size of course) {

            if (sorted.length >= size) {
                combination(sorted, size, 0, "");
            }
        }
    }

    const answer = [];

    for (const size of course) {

        let max = 1;
        let result = [];

        for (const [key, value] of map) {

            if (key.length !== size) continue;

            if (value === max && max > 1) {
                result.push(key);
            } else if (value > max) {
                max = value;
                result = [key];
            }
        }

        answer.push(...result);
    }

    return answer.sort();
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;

class Solution {

    Map<String, Integer> map = new HashMap<>();

    public String[] solution(String[] orders, int[] course) {

        for (String order : orders) {

            char[] chars = order.toCharArray();
            Arrays.sort(chars);

            for (int size : course) {

                if (chars.length >= size) {
                    combination(chars, size, 0, new StringBuilder());
                }
            }
        }

        List<String> answer = new ArrayList<>();

        for (int size : course) {

            int max = 1;
            List<String> result = new ArrayList<>();

            for (String key : map.keySet()) {

                if (key.length() != size) continue;

                int cnt = map.get(key);

                if (cnt == max && max > 1) {
                    result.add(key);
                } else if (cnt > max) {
                    max = cnt;
                    result.clear();
                    result.add(key);
                }
            }

            answer.addAll(result);
        }

        Collections.sort(answer);

        return answer.toArray(new String[0]);
    }

    void combination(char[] arr, int target, int idx, StringBuilder sb) {

        if (sb.length() == target) {

            String key = sb.toString();

            map.put(key, map.getOrDefault(key, 0) + 1);

            return;
        }

        for (int i = idx; i < arr.length; i++) {

            sb.append(arr[i]);

            combination(arr, target, i + 1, sb);

            sb.deleteCharAt(sb.length() - 1);
        }
    }

}